### PR TITLE
Switch to using informers for all awaits

### LINKED ISSUE #1598: Restructure await logic to use shared informers instead of relying on fine-grained watches
Currently our await logic relies on setting up independent watches on each resource (and in some cases for higher order resources like deployments/stateful sets etc., on their respective components like pods, replicasets, PVCs etc.). This can cause a lot of redundant information to be spewed to the watches. We should restructure the awaits to use a [shared informer](https://pkg.go.dev/k8s.io/client-go/informers#SharedInformerFactory) on the namespace witch filter options.

This will also necessitate less reliance on strong consistency/ordering of events which should make the await logic less brittle (e.g. see https://github.com/pulumi/pulumi-kubernetes/issues/1502)

### PR DESCRIPTION
**Pull request by viveklak**
Fixes #1598

### CODE REVIEW

**Comment by viveklak** [comment]
@lblackstone this is ready for an initial review. I intend to add another couple of tests. Specifically would like to include a test to ensure ingress/endpoints await works as desired but manual testing suggested this was good.

The only place left with a direct call to watch is the generic delete handler here: https://github.com/pulumi/pulumi-kubernetes/blob/master/provider/pkg/await/await.go#L474. I will look to get rid of this as well.

File: provider/pkg/await/service.go [hunk 1]
```diff
@@ -120,32 +122,34 @@ func (sia *serviceInitAwaiter) Await() error {
...
+	go endpointsInformer.Informer().Run(stopper)
 
 	version := cluster.TryGetServerVersion(sia.config.clientSet.DiscoveryClientCached)
 
 	timeout := metadata.TimeoutDuration(sia.config.timeout, sia.config.currentInputs, DefaultServiceTimeoutMins*60)
-	return sia.await(serviceWatcher, endpointWatcher, time.After(timeout), make(chan struct{}), version)
+	return sia.await(serviceEvents, endpointsEvents, time.After(timeout), make(chan struct{}), version)
 }
 
 func (sia *serviceInitAwaiter) Read() error {
```
Comment (lblackstone) [inline provider/pkg/await/service.go:155]: I noticed that the Job awaiter Read method was updated to use the Informer.Logger. Any reason to prefer the old way over that?

File: provider/pkg/await/service.go [hunk 2]
```diff
@@ -216,7 +220,8 @@ func (sia *serviceInitAwaiter) read(
 
 // await is a helper companion to `Await` designed to make it easy to test this module.
 func (sia *serviceInitAwaiter) await(
-	serviceWatcher, endpointWatcher watch.Interface,
+	serviceEvents <-chan watch.Event,
+	endpointsEvents <-chan watch.Event,
```
Comment (lblackstone) [inline provider/pkg/await/service.go:224]:
```suggestion
	serviceEvents, endpointsEvents <-chan watch.Event,
```

**Review by lblackstone — COMMENTED** [review]
Looks good overall

File: provider/pkg/await/service.go (hunk 1 above)
Comment (viveklak) [reply provider/pkg/await/service.go:155]: The job awaiter's Read uses PodAggregator which I thought would be a good candidate for using a lister backed by the informer anyway (as part of the corresponding watch). As a result, I switched Job awaiter's Read to use an informer as well. For the other awaiters I didn't see any advantage to switching Read to informers so let that be.

**Review by lblackstone — APPROVED** [review]

**Comment by viveklak** [comment]
The one CI failure should be fixed when 3.7.0 is used.

**Pull request merged**